warp-tls/Network/Wai/Handler/WarpTLS.hs

```diff
@@ -71,17 +71,29 @@ import System.IO.Error (isEOFError, ioeGetErrorType)
 
 ----------------------------------------------------------------
 
+-- | Determines where to load the certificate, chain 
+-- certificates, and key from.
+data CertSettings 
+  = CertFromFile !FilePath ![FilePath] !FilePath
+  | CertFromMemory !S.ByteString ![S.ByteString] !S.ByteString
+  | CertFromRef !(I.IORef S.ByteString) ![I.IORef S.ByteString] !(I.IORef S.ByteString)
```

Comment:
I'm rather late to party, sorry about that, just noticed this PR today, and sadly it has been merged for some time now.  And yet perhaps I should make my comments anyway.  My first issue is with the choice of separate IORefs for each of the cert, chain and keys.  I think that's likely a mistake.  Rather I think that the IORef representation should have been:
```haskell
CertFromRef !(I.IORef (S.ByteString,  [S.ByteString], S.ByteString))
```
That is, a single `IORef` holding a triple with the certificate, issuer chain, and key.  This would allow atomic retrieval of the triple, in which one can be confident that all three objects correspond to each other.  The current design is subject race conditions between reader and writer, because the writer updates multiple `IORef`s separately, and the reader may see an inconsistent view of the data. :-(